### PR TITLE
Yul dataflow analysis for continue/break statements

### PR DESCRIPTION
**Pull request by christianparpart**
Adds data flow analysis for break & continue statements. This PR is **rebased** on top of #6136.

### CODE REVIEW

File: libyul/optimiser/DataFlowAnalyzer.cpp [hunk 1]
```diff
@@ -117,27 +117,29 @@ void DataFlowAnalyzer::operator()(ForLoop& _for)
...
+
 	Assignments assignments;
 	assignments(_for.body);
 	assignments(_for.post);
 	clearValues(assignments.names());
 
 	visit(*_for.condition);
 	(*this)(_for.body);
+	if (m_assignmentsSinceCont)
+		clearValues(m_assignmentsSinceCont->names());
```
Comment (leonardoalt) [inline libyul/optimiser/DataFlowAnalyzer.cpp]: Won't this be always empty?

Comment (leonardoalt) [reply libyul/optimiser/DataFlowAnalyzer.cpp]: Since `m_assignmentsSinceCont` isn't called on `Assignment`s

Comment (christianparpart) [reply libyul/optimiser/DataFlowAnalyzer.cpp]: It's being updated in `operator()(Continue&)`'s body.

Comment (leonardoalt) [reply libyul/optimiser/DataFlowAnalyzer.cpp]: Yes, but why isn't it still empty? Shouldn't it call `m_assignmentsSinceCont(_assignment)` for `Assignment`?

Comment (christianparpart) [reply libyul/optimiser/DataFlowAnalyzer.cpp]: Ah, now I believe I got your point! Of course, but then I've to rewrite the code quite a bit. Hold on. :)

File: libyul/optimiser/NameCollector.h [hunk 2]
```diff
@@ -81,4 +81,28 @@ class Assignments: public ASTWalker
 	std::set<YulString> m_names;
 };
 
+/**
+ * Collects all names from a given continue statement on onwards.
```
Comment (chriseth) [inline libyul/optimiser/NameCollector.h:85]:
```suggestion
 * Collects all names from a given continue statement onwards.
```

File: libyul/optimiser/NameCollector.h [hunk 3]
```diff
@@ -81,4 +81,28 @@ class Assignments: public ASTWalker
...
+ *
+ * It makes only sense to be invoked from within a body of an outer for loop, that is,
```
Comment (chriseth) [inline libyul/optimiser/NameCollector.h:87]:
```suggestion
 * It makes only sense to be invoked on a body of an outer for loop, that is,
```
?

File: libyul/optimiser/NameCollector.h [hunk 4]
```diff
@@ -81,4 +81,28 @@ class Assignments: public ASTWalker
...
+ * it will only collect all names from the beginning of the first continue statement
+ * of the outer-most ForLoop.
```
Comment (chriseth) [inline libyul/optimiser/NameCollector.h:89]: ... and ignores all for loops it encounters.

File: libyul/optimiser/NameCollector.h [hunk 5]
```diff
@@ -81,4 +81,28 @@ class Assignments: public ASTWalker
...
+ */
+class AssignmentsSinceContinue: public ASTWalker
+{
+public:
+	using ASTWalker::operator();
+	void operator()(ForLoop const& _forLoop) override;
+	void operator()(Continue const&) override;
+	void operator()(Assignment const& _assignment) override;
+
```
Comment (chriseth) [inline libyul/optimiser/NameCollector.h:99]: Should ignore functions

File: test/libyul/yulOptimizerTests/commonSubexpressionEliminator/for_continue.yul [hunk 6]
```diff
@@ -0,0 +1,38 @@
+{
```
Comment (leonardoalt) [inline test/libyul/yulOptimizerTests/commonSubexpressionEliminator/for_continue.yul]: What is this test testing?

**Review by leonardoalt — COMMENTED** [review]
Is this WIP now? I see some come vanished :p

**Comment by chriseth** [comment]
The code vanished because it turned out that we actually do not have to do anything. Would be great if you could verify that, @leonardoalt

**Comment by leonardoalt** [comment]
Ok I see. It would be nice to have tests showing that.

File: test/libyul/yulOptimizerTests/rematerialiser/for_break.yul [hunk 7]
```diff
@@ -0,0 +1,30 @@
+{
+    let a := caller()
+	let k := 1
```
Comment (chriseth) [inline test/libyul/yulOptimizerTests/rematerialiser/for_break.yul]: Indentation

File: test/libyul/yulOptimizerTests/rematerialiser/for_break.yul [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    for {let i := 0} lt(i, 10) { i := add(i, k) } {
```
Comment (chriseth) [inline test/libyul/yulOptimizerTests/rematerialiser/for_break.yul]:
```
  let a
  let b
  for {let i := 0} lt(i, 10) { i := add(a, b) } {
    a := origin()
    b := origin()
    b := caller()
    if callvalue() { break }
    a := caller()
}
````

File: test/libyul/yulOptimizerTests/rematerialiser/for_break.yul [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+// rematerialiser
+// {
+//     let a
+//     let b
+//     for {
+//         let i := 0
+//     }
+//     lt(i, 10)
+//     {
+//         i := add(caller(), caller())
```
Comment (chriseth) [inline test/libyul/yulOptimizerTests/rematerialiser/for_break.yul:25]: This should be `i := add(a, caller())`

Comment (chriseth) [reply test/libyul/yulOptimizerTests/rematerialiser/for_break.yul:25]: Sorry - it's correct here, but the continue test should be like this.

File: libyul/optimiser/NameCollector.h (hunk 5 above)
Comment (chriseth) [inline libyul/optimiser/NameCollector.h:99]: Please add `operator()(FunctionDefinition)` and assert if it is called.

File: libyul/optimiser/NameCollector.h [hunk 10]
```diff
@@ -81,4 +81,28 @@ class Assignments: public ASTWalker
...
+	std::set<YulString> const& names() const { return m_names; }
+	bool empty() const noexcept { return m_names.empty(); }
+
+private:
+	int m_forLoopDepth = 0;
```
Comment (chriseth) [inline libyul/optimiser/NameCollector.h]: `size_t`?

File: libyul/optimiser/DataFlowAnalyzer.cpp [hunk 11]
```diff
@@ -117,27 +117,26 @@ void DataFlowAnalyzer::operator()(ForLoop& _for)
...
+
+	if (!assignmentsSinceCont.empty())
+		clearValues(assignmentsSinceCont.names());
+
 	(*this)(_for.post);
 
 	clearValues(assignments.names());
-	popScope();
-}
+	// TODO: test: continue + assignment inside post{}
```
Comment (chriseth) [inline libyul/optimiser/DataFlowAnalyzer.cpp]: Is that still pending?

File: libyul/optimiser/DataFlowAnalyzer.cpp [hunk 12]
```diff
@@ -117,27 +117,26 @@ void DataFlowAnalyzer::operator()(ForLoop& _for)
...
+
 	Assignments assignments;
 	assignments(_for.body);
 	assignments(_for.post);
 	clearValues(assignments.names());
 
 	visit(*_for.condition);
 	(*this)(_for.body);
+
+	if (!assignmentsSinceCont.empty())
```
Comment (chriseth) [inline libyul/optimiser/DataFlowAnalyzer.cpp]: Is the empty check needed?

File: test/libyul/yulOptimizerTests/rematerialiser/for_break.yul [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+	let b
+	for {let i := 0} lt(i, 10) {i := add(a, b)} {
+		a := origin()
+		b := origin()
+
+		b := caller()
+		if callvalue() { break }
+		a := caller()
+	}
+}
```
Comment (chriseth) [inline test/libyul/yulOptimizerTests/rematerialiser/for_break.yul:14]: Actually we could do an `mstore(a, b)` here just to check that it is cleared after the loop.

File: test/libyul/yulOptimizerTests/rematerialiser/for_continue.yul [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+	for {let i := 0} lt(i, 10) {i := add(a, b)}
+	{
+		a := origin()
+		b := origin()
+
+		b := caller()
+		if callvalue() { continue }
+		a := caller()
+	}
+}
```
Comment (chriseth) [inline test/libyul/yulOptimizerTests/rematerialiser/for_continue.yul:16]: Same here.

File: test/libyul/yulOptimizerTests/rematerialiser/for_continue.yul [hunk 15]
```diff
@@ -0,0 +1,36 @@
+{
+	let a
+	let b
+	for {let i := 0} lt(i, 10) {i := add(a, b)}
```
Comment (chriseth) [inline test/libyul/yulOptimizerTests/rematerialiser/for_continue.yul]: Perhaps add a comment: `a` might be reset after the continue, so we can only replace `b` which is known to be `caller()`.

File: test/libyul/yulOptimizerTests/rematerialiser/for_continue_with_assignment_in_post.yul [hunk 16]
```diff
@@ -0,0 +1,39 @@
+{
+	let a
+	let b
+	for {let i := 0} lt(i, 10) {i := add(a, b) b := origin()}
```
Comment (chriseth) [inline test/libyul/yulOptimizerTests/rematerialiser/for_continue_with_assignment_in_post.yul]: Hm, unfortunately, I don't remember anymore which aspect this was supposed to test..

File: libyul/optimiser/DataFlowAnalyzer.cpp (hunk 11 above)
Comment (christianparpart) [reply libyul/optimiser/DataFlowAnalyzer.cpp]: ah, you're right. I accidentally forgot to remove my personal note here.

File: test/libyul/yulOptimizerTests/rematerialiser/for_continue_with_assignment_in_post.yul (hunk 16 above)
Comment (christianparpart) [reply test/libyul/yulOptimizerTests/rematerialiser/for_continue_with_assignment_in_post.yul]: @chriseth the motivation behind this, yesterday, was, to test for loops with continue statements modifying a used variable in the post-block (`b := origin()`). Maybe I should have chosen a different value?

File: test/libyul/yulOptimizerTests/rematerialiser/for_continue_with_assignment_in_post.yul [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+//     for {
+//         let i := 0
+//         b := origin()
+//         c := origin()
+//     }
+//     lt(i, 10)
+//     {
+//         i := add(a, caller())
+//         b := 0x42
+//         c := 0x90
```
Comment (chriseth) [inline test/libyul/yulOptimizerTests/rematerialiser/for_continue_with_assignment_in_post.yul]: Ok, final comment: Please do not use constants, they might be more expensive than e.g. `caller()` and thus not rematerialize.

Comment (christianparpart) [reply test/libyul/yulOptimizerTests/rematerialiser/for_continue_with_assignment_in_post.yul]: didn't know constants are more expensive. I just didn't wanna use caller() everywhere. :)

Comment (chriseth) [reply test/libyul/yulOptimizerTests/rematerialiser/for_continue_with_assignment_in_post.yul]: Look for a cost of `BASE` in `Instruction.cpp`

Comment (christianparpart) [reply test/libyul/yulOptimizerTests/rematerialiser/for_continue_with_assignment_in_post.yul]: I've replaced the constants with other caller() / callvalue().

**Review by chriseth — APPROVED** [review]

**Pull request merged**